apps/studio/electron/main/run/setup.ts

```diff
@@ -0,0 +1,114 @@
+import traverse, { NodePath } from '@babel/traverse';
+import * as t from '@babel/types';
+import { EditorAttributes } from '@onlook/models/constants';
+import type { TemplateNode } from '@onlook/models/element';
+import { generateCode } from '../code/diff/helpers';
+import { formatContent, readFile } from '../code/files';
+import { parseJsxFile } from '../code/helpers';
+import { generateCodeOptions, generateId, getTemplateNode, isReactFragment } from './helpers';
+
+export async function getFileWithIds(filePath: string): Promise<string | null> {
+    const content = await readFile(filePath);
+    const ast = parseJsxFile(content);
+    if (!ast) {
+        console.error(`Failed to parse file: ${filePath}`);
+        return null;
+    }
+    addIdsToAst(ast);
+    const generated = generateCode(ast, generateCodeOptions, content);
+    const formatted = await formatContent(filePath, generated);
+    return formatted;
+}
+
+function addIdsToAst(ast: t.File) {
+    const ids: Set<string> = new Set();
+
+    traverse(ast, {
+        JSXOpeningElement(path: NodePath<t.JSXOpeningElement>) {
+            if (isReactFragment(path.node)) {
+                return;
+            }
+            const attributes = path.node.attributes;
+            const existingAttrIndex = attributes.findIndex(
+                (attr: any) => attr.name?.name === EditorAttributes.DATA_ONLOOK_ID,
+            );
+
+            if (existingAttrIndex !== -1) {
+                const existingId = (attributes[existingAttrIndex] as any).value.value;
+                if (ids.has(existingId)) {
+                    const newId = generateId();
+                    (attributes[existingAttrIndex] as any).value.value = newId;
+                    ids.add(newId);
+                } else {
+                    ids.add(existingId);
+                }
+                return;
+            }
+
+            const elementId = generateId();
+            const oid = t.jSXAttribute(
+                t.jSXIdentifier(EditorAttributes.DATA_ONLOOK_ID),
+                t.stringLiteral(elementId),
+            );
+            attributes.push(oid);
+            ids.add(elementId);
+        },
+    });
+}
+
+export function createMappingFromContent(content: string, filename: string) {
+    const ast = parseJsxFile(content);
+    if (!ast) {
+        return null;
+    }
+    return createMapping(ast, filename);
+}
+
+function createMapping(ast: t.File, filename: string): Record<string, TemplateNode> | null {
```

Comment:
@iNerdStack , fyi. This is where information about the code is being computed. We can extend TemplateNode in order to add data such as array item. See the componentStack which does something similar by keeping a stack of the scope name